File: 1793-maximum-score-of-a-good-subarray/1793-maximum-score-of-a-good-subarray.py
class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        n = len(nums)
        left = [-1] * n
        stack = []
        
        for i in range(n - 1, -1, -1):
            while stack and nums[stack[-1]] > nums[i]:
                left[stack.pop()] = i
            stack.append(i)
        
        right = [n] * n
        stack = []
        
        for i in range(n):
            while stack and nums[stack[-1]] > nums[i]:
                right[stack.pop()] = i
            stack.append(i)
        
        ans = 0
        for i in range(n):
            if left[i] < k and right[i] > k:
                ans = max(ans, nums[i] * (right[i] - left[i] - 1))
        return ans
        

File: 1793-maximum-score-of-a-good-subarray/1793-maximum-score-of-a-good-subarray.cpp
class Solution {
public:
    int maximumScore(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> left(n, -1);
        stack<int> st;
        for (int i = n - 1; i >= 0; --i){
            while (!st.empty() && nums[st.top()] > nums[i]){
                left[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }
        
        vector<int> right(n, n);
        st = stack<int>();
        for (int i = 0; i < n; ++i){
            while (!st.empty() && nums[st.top()] > nums[i]){
                right[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }
        
        int ans = 0;
        for (int i = 0; i < n; ++i){
            if (left[i] < k && right[i] > k){
                ans = max(ans, nums[i] * (right[i] - left[i] - 1));
            }
        }
        
        return ans;
    }
};